docs/internals/optimizer.rst

```diff
@@ -0,0 +1,1263 @@
+of this form by just swapping out the declaration and the assignment. The above
+snippet is turned into
+
+::
+
+    a := E
+    let a_1 := a
+    mstore(a_1, 1)
+
+This is a very simple equivalence transform, but when we now run the
+Common Subexpression Eliminator, it will replace all occurrences of ``a_1``
+by ``a`` (until ``a`` is re-assigned). The Unused Pruner will then
+eliminate the variable ``a_1`` altogether and thus fully reverse the
+SSA transform.
+
+.. _stack-compressor:
+
+StackCompressor
+^^^^^^^^^^^^^^^
+
+One problem that makes code generation for the Ethereum Virtual Machine
+hard is the fact that there is a hard limit of 16 slots for reaching
+down the expression stack. This more or less translates to a limit
+of 16 local variables. The stack compressor takes Yul code and
+compiles it to EVM bytecode. Whenever the stack difference is too
+large, it records the function this happened in.
+
+For each function that caused such a problem, the Rematerialiser
+is called with a special request to aggressively eliminate specific
+variables sorted by the cost of their values.
+
+On failure, this procedure is repeated multiple times.
+
+.. _rematerialiser:
+
+Rematerialiser
+^^^^^^^^^^^^^^
+
+The rematerialisation stage tries to replace variable references by the expression that
+was last assigned to the variable. This is of course only beneficial if this expression
+is comparatively cheap to evaluate. Furthermore, it is only semantically equivalent if
+the value of the expression did not change between the point of assignment and the
+point of use. The main benefit of this stage is that it can save stack slots if it
+leads to a variable being eliminated completely (see below), but it can also
+save a DUP opcode on the EVM if the expression is very cheap.
+
+The Rematerialiser uses the Dataflow Analyzer to track the current values of variables,
+which are always movable.
+If the value is very cheap or the variable was explicitly requested to be eliminated,
+the variable reference is replaced by its current value.
+
+.. _for-loop-condition-out-of-body:
+
+ForLoopConditionOutOfBody
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Reverses the transformation of ForLoopConditionIntoBody.
+
+For any movable ``c``, it turns
+
+::
+
+    for { ... } 1 { ... } {
+    if iszero(c) { break }
+    ...
+    }
+
+into
+
+::
+
+    for { ... } c { ... } {
+    ...
+    }
+
+and it turns
+
+::
+
+    for { ... } 1 { ... } {
+    if c { break }
+    ...
+    }
+
+into
+
+::
+
+    for { ... } iszero(c) { ... } {
+    ...
+    }
+
+The LiteralRematerialiser should be run before this step.
+
+
+WebAssembly specific
+--------------------
+
+MainFunction
+^^^^^^^^^^^^
+
+Changes the topmost block to be a function with a specific name ("main") which has no
+inputs nor outputs.
+
+Depends on the Function Grouper.
+
+## Below steps still need to be sorted into the list above according to their category.
+
+.. _expression-inliner:
+
+ExpressionInliner
```

Comment:
This is the same as `FunctionalInliner`. Can you combine them? I think the second point in the second list here is not covered above.